### DISCUSSION TITLE
Pre-set default values

### DISCUSSION THREAD

**Discussion by Venustiano** [rubenarslan/formr.org#534] [Q&A]
From the `Formr` documentation I can see that we can create a 'value' column to pre-set 'default' values using R code as follows
```
 a_different_survey$item1 + a_different_survey$item2 
```
If I understand correctly, to be able to use such a technique, `a_different_survey` must be in the `Formr` server. In our case, this is not the case. The survey data is in excel/google sheets. So, I have a couple of questions. 

1. Is there a way to populate `a_different_survey` programmatically in `Formr`?
2. Is it possible to **pre-set** values using data directly from **excel** files?

Thanks in advance

Venus

**Comment by dfeistauer** [comment]
Hi Venus,

write your R-Code in the column "value" in Excel. 
Sometimes you need to use as.numeric()
between item2 and the pre-set-Variable needs to be a submit-Button.

Daniela

**Reply by Venustiano** [reply]
Indeed, I had to use `as.numeric()`
Thanks for the advice

**Comment by Venustiano** [comment]
How do pre-set a value when the type is `select_one`?
In the `R` source code, the selected value is OK, but in the form the value is not selected as in the screenshot below. `Gegevens gemeente (prev)` should be one of the options in the drop down `menu`.

![image](https://github.com/rubenarslan/formr.org/assets/39260912/7cad36ba-29de-4c87-ba68-544fbef9276b)

**Reply by dfeistauer** [reply]
please send a screenshot of your excel sheet

**Reply by Venustiano** [reply]
Thank you for your quick reply. 

Thanks to your request I realized that I did not capitalize the `name` column. 

![image](https://github.com/rubenarslan/formr.org/assets/39260912/246348c2-8447-44b0-b7a7-fd0afd51de93)

Now it works fine.

**Reply by dfeistauer** [reply]
I meant the excel with the value column.

**Reply by Venustiano** [reply]
The value column is like this

![image](https://github.com/rubenarslan/formr.org/assets/39260912/6d47be1c-9fd0-4a6e-9066-0c01b98e78df)

But the problem was that values in the data are capitalized but my `name` column was not. I had something like `groningen` and `Groningen`

**Reply by dfeistauer** [reply]
great

**Comment by Venustiano** [comment]
Hello again,

Is there a way to create 'persistent' variable across questions. For example, in the screenshot below the variable `gname_prev`, in row 4 which is loaded from googledrive, cannot be reused in row 6. 

<img width="1364" alt="image" src="https://github.com/rubenarslan/formr.org/assets/39260912/b98f2d33-1732-4171-992b-97de02cd9b22">

It leads to the following error.

```
R Error: object 'gname_prev' not found

In call:
(function() {
    previous_value <- gname_prev[which(as.numeric(gname_prev$rangnr) == 
        id_prev_survey), "Algtelefoonnummer"]
    previous_value
})()
```
 To fix the error I have to load the data again

<img width="1361" alt="image" src="https://github.com/rubenarslan/formr.org/assets/39260912/b8675e98-746d-445a-9942-504ab8687802">


Is there a way to reuse variables in `Formr`?

Thanks in advance

**Reply by rubenarslan** [reply]
Maybe this example helps: https://docs.google.com/spreadsheets/d/1jkts0xp1JmFX5t1A34kLswH7ZyWWn8xArKeCH2-pzH4/edit#gid=2037039862

You could also set a global variable using `gname_prev <<- rio::etc`, I haven't tested this.

**Reply by Venustiano** [reply]
Thank you for your suggestions @rubenarslan,

The global variable works in a single page, moving to the next page, the variable has to be loaded again. It is better than loading the variable per question. However, it would be much better if this could work across pages. 

Regarding the google sheet example,

```
library(ggplot2)
data(diamonds)
selected <- diamonds[sample(nrow(diamonds),size = 180) ,]
jsonlite::toJSON(selected)
```

 it is not clear to me how the variable `selected` could be reused without loading the data again. At least I can not see it in the example. Am I missing something?

**Reply by rubenarslan** [reply]
Hi @Venustiano maybe provide some more information on your use case. In the example, the R variable "selected" is stored as JSON in the variable "item_list" in formr. You're correct that R variables do not persist across page loads.

**Reply by Venustiano** [reply]
Hi @rubenarslan 

Basically, I need to create multiple pages. For each page, I need to read data (from a google sheet) from previous surveys. However, reading/loading data for each page is very slow. Reading the data only once will improve user experience. It is also better during the testing phase. 

I have noticed that a `type number` variable persists across pages. However, a `type hidden` does not. For example, in the screenshot below, `id_prev_survey` persist across all pages, but `gname_prev` which stores `google sheet` persist only in the same page. For the next page I have to load the data again, otherwise I get an error.

![image](https://github.com/rubenarslan/formr.org/assets/39260912/4a4cccd4-5272-4c9c-991b-bc14f67b07cc)

Do you know how to make `gname_prev` persistent across pages?

Thanks
